libraries/hotstuff/include/eosio/hotstuff/qc_chain.hpp

```diff
@@ -134,31 +103,45 @@ namespace eosio::hotstuff {
       void send_hs_new_view_msg(const hs_new_view_message& msg); //send new view msg
       void send_hs_new_block_msg(const hs_new_block_message& msg); //send new block msg
 
-      void on_hs_vote_msg(const hs_vote_message& msg); //vote msg event handler
-      void on_hs_proposal_msg(const hs_proposal_message& msg); //proposal msg event handler
-      void on_hs_new_view_msg(const hs_new_view_message& msg); //new view msg event handler
-      void on_hs_new_block_msg(const hs_new_block_message& msg); //new block msg event handler
-
       void update(const hs_proposal_message& proposal); //update internal state
       void commit(const hs_proposal_message& proposal); //commit proposal (finality)
 
       void gc_proposals(uint64_t cutoff); //garbage collection of old proposals
 
-private:
-
-      // This mutex synchronizes all writes to the data members of this qc_chain against
-      //   get_state() calls (which ultimately come from e.g. the HTTP plugin).
-      // This could result in a HTTP query that gets the state of the core while it is
-      //   in the middle of processing a given request, since this is not serializing
-      //   against high-level message or request processing borders.
-      // If that behavior is not desired, we can instead synchronize this against a
-      //   consistent past snapshot of the qc_chain's state for e.g. the HTTP plugin,
-      //   which would be updated at the end of processing every request to the core
-      //   that does alter the qc_chain (hotstuff protocol state).
-      // And if the chain_pacemaker::_hotstuff_global_mutex locking strategy is ever
-      //   changed, then this probably needs to be reviewed as well.
-      //
-      mutable std::mutex     _state_mutex;
+#warning remove. bls12-381 key used for testing purposes
+      //todo : remove. bls12-381 key used for testing purposes
+      std::vector<uint8_t> _seed =
+         {  0,  50, 6,  244, 24,  199, 1,  25,  52,  88,  192,
+            19, 18, 12, 89,  6,   220, 18, 102, 58,  209, 82,
+            12, 62, 89, 110, 182, 9,   44, 20,  254, 22 };
+
+      fc::crypto::blslib::bls_private_key _private_key = fc::crypto::blslib::bls_private_key(_seed);
+
+      enum msg_type {
+         new_view = 1,
+         new_block = 2,
+         qc = 3,
+         vote = 4
+      };
+
+      bool _chained_mode = false;
+      block_id_type _block_exec = NULL_BLOCK_ID;
+      block_id_type _pending_proposal_block = NULL_BLOCK_ID;
+      fc::sha256 _b_leaf = NULL_PROPOSAL_ID;
+      fc::sha256 _b_lock = NULL_PROPOSAL_ID;
+      fc::sha256 _b_exec = NULL_PROPOSAL_ID;
+      fc::sha256 _b_finality_violation = NULL_PROPOSAL_ID;
+      eosio::chain::quorum_certificate _high_qc;
+      eosio::chain::quorum_certificate _current_qc;
+      uint32_t _v_height = 0;
+      eosio::chain::extended_schedule _schedule;
```

Comment:
Why do we duplicate almost all members of `struct finalizer_state` here? It would be better to have the same struct declaration here (maybe creating a base struct for `finalizer_state` with the common members (without the `std::map`). This would also make the `qc_chain::get_state(finalizer_state& fs)` simpler.

Reply:
`finalizer_state` was created for the `get_finalizer_state` API. I can create a `finalizer_state  _fs` member of `qc_chain`, but it will leave somethings out since not all state is relevant for the API.

Reply:
I think we may find a somewhat cleaner way to do this eventually, but it is not critical and I don't want to delay the review.

Reply:
`finalizer_state` ended up being useful for the tests as well (once we moved all the qc_chain member vars to private).